modules/lang-painless/src/main/java/org/elasticsearch/painless/Scope.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless;
+
+import org.elasticsearch.painless.lookup.PainlessLookupUtility;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Tracks user defined variables across compilation phases.
+ */
+public abstract class Scope {
+
+    public static class Variable {
+
+        protected final Class<?> type;
+        protected final String name;
+        protected final boolean isReadOnly;
+
+        public Variable(Class<?> type, String name, boolean isReadOnly) {
+            this.type = Objects.requireNonNull(type);
+            this.name = Objects.requireNonNull(name);
+            this.isReadOnly = isReadOnly;
+        }
+
+        public Class<?> getType() {
+            return type;
+        }
+
+        public String getCanonicalTypeName() {
+            return PainlessLookupUtility.typeToCanonicalTypeName(type);
+        }
+
+        public String getName() {
+            return name;
+        }
+
+        public boolean isReadOnly() {
+            return isReadOnly;
+        }
+    }
+
+    public static class FunctionScope extends Scope {
+
+        protected final Class<?> returnType;
+        protected final Set<String> areReadFrom = new HashSet<>();
```

Comment:
Maybe we could use "used" terminology instead of "read from", as the latter is difficult to read (is it "reed" or "red"). I would also just have it called "used" or "usedVariables", not "areUsed" since "are" and "is" imply the type should be boolean. Also can you please add a comment explaining the semantics (eg used means getVariable was called for that name).

Reply:
Agreed and changed.